Index: Objects/exceptions.c
===================================================================
--- Objects/exceptions.c	(revision 82153)
+++ Objects/exceptions.c	(working copy)
@@ -1784,9 +1784,6 @@
     const char *encoding, const char *object, Py_ssize_t length,
     Py_ssize_t start, Py_ssize_t end, const char *reason)
 {
-    assert(length < INT_MAX);
-    assert(start < INT_MAX);
-    assert(end < INT_MAX);
     return PyObject_CallFunction(PyExc_UnicodeDecodeError, "ss#nns",
                                  encoding, object, length, start, end, reason);
 }
