--- Python-2.3.4/Lib/pydoc.py.no-doc	2004-07-16 11:29:01.000000000 -0400
+++ Python-2.3.4/Lib/pydoc.py	2004-07-16 11:32:52.000000000 -0400
@@ -1524,6 +1524,7 @@
         homedir = os.environ.get('PYTHONHOME')
         for dir in [os.environ.get('PYTHONDOCS'),
                     homedir and os.path.join(homedir, 'doc'),
+                    '/usr/share/doc/python-docs-%s/html' % split(sys.version)[0],
                     os.path.join(execdir, 'doc'),
                     '/usr/doc/python-docs-' + split(sys.version)[0],
                     '/usr/doc/python-' + split(sys.version)[0],
