diff -up Python-2.5.1/Modules/_ctypes/libffi/src/x86/unix64.S.execstack Python-2.5.1/Modules/_ctypes/libffi/src/x86/unix64.S
--- Python-2.5.1/Modules/_ctypes/libffi/src/x86/unix64.S.execstack	2007-08-10 10:34:06.000000000 +0200
+++ Python-2.5.1/Modules/_ctypes/libffi/src/x86/unix64.S	2007-08-10 10:34:06.000000000 +0200
@@ -410,3 +410,7 @@ ffi_closure_unix64:
 .LEFDE3:
 
 #endif /* __x86_64__ */
+
+#ifdef __ELF__
+.section .note.GNU-stack,"",%progbits
+#endif
diff -up Python-2.5.1/Modules/_ctypes/libffi/src/powerpc/sysv.S.execstack Python-2.5.1/Modules/_ctypes/libffi/src/powerpc/sysv.S
--- Python-2.5.1/Modules/_ctypes/libffi/src/powerpc/sysv.S.execstack	2007-08-10 10:39:03.000000000 +0200
+++ Python-2.5.1/Modules/_ctypes/libffi/src/powerpc/sysv.S	2007-08-10 10:39:59.000000000 +0200
@@ -215,3 +215,7 @@ END(ffi_call_SYSV)
       .align 2
 .LEFDE1:
 #endif
+
+#ifdef __ELF__
+.section .note.GNU-stack,"",%progbits
+#endif
diff -up Python-2.5.1/Modules/_ctypes/libffi/src/powerpc/linux64_closure.S.execstack Python-2.5.1/Modules/_ctypes/libffi/src/powerpc/linux64_closure.S
--- Python-2.5.1/Modules/_ctypes/libffi/src/powerpc/linux64_closure.S.execstack	2007-08-10 10:39:23.000000000 +0200
+++ Python-2.5.1/Modules/_ctypes/libffi/src/powerpc/linux64_closure.S	2007-08-10 10:39:58.000000000 +0200
@@ -204,3 +204,7 @@ ffi_closure_LINUX64:
 	.align 3
 .LEFDE1:
 #endif
+
+#ifdef __ELF__
+.section .note.GNU-stack,"",%progbits
+#endif
diff -up Python-2.5.1/Modules/_ctypes/libffi/src/powerpc/linux64.S.execstack Python-2.5.1/Modules/_ctypes/libffi/src/powerpc/linux64.S
--- Python-2.5.1/Modules/_ctypes/libffi/src/powerpc/linux64.S.execstack	2007-08-10 10:39:13.000000000 +0200
+++ Python-2.5.1/Modules/_ctypes/libffi/src/powerpc/linux64.S	2007-08-10 10:40:01.000000000 +0200
@@ -178,3 +178,7 @@ ffi_call_LINUX64:
 	.align 3
 .LEFDE1:
 #endif
+
+#ifdef __ELF__
+.section .note.GNU-stack,"",%progbits
+#endif
