--- Python-2.5/Modules/_ctypes/libffi/src/x86/unix64.S.ctypes-exec	2006-03-08 20:35:32.000000000 -0500
+++ Python-2.5/Modules/_ctypes/libffi/src/x86/unix64.S	2007-01-03 12:48:32.000000000 -0500
@@ -410,3 +410,5 @@
 .LEFDE3:
 
 #endif /* __x86_64__ */
+
+.section .note.GNU-stack,"",@progbits
--- Python-2.5/Modules/_ctypes/libffi/src/x86/sysv.S.ctypes-exec	2006-03-08 20:35:32.000000000 -0500
+++ Python-2.5/Modules/_ctypes/libffi/src/x86/sysv.S	2007-01-03 12:48:32.000000000 -0500
@@ -376,3 +376,5 @@
 #endif
 
 #endif /* ifndef __x86_64__ */
+
+.section .note.GNU-stack,"",@progbits
--- Python-2.5/Modules/_ctypes/libffi/src/s390/sysv.S.ctypes-exec	2007-01-03 12:54:40.000000000 -0500
+++ Python-2.5/Modules/_ctypes/libffi/src/s390/sysv.S	2007-01-03 12:51:32.000000000 -0500
@@ -427,3 +427,4 @@
 
 #endif
 
+.section .note.GNU-stack,"",@progbits
--- Python-2.5/Modules/_ctypes/libffi/src/powerpc/linux64.S.ctypes-exec	2007-01-03 12:54:40.000000000 -0500
+++ Python-2.5/Modules/_ctypes/libffi/src/powerpc/linux64.S	2007-01-03 12:51:51.000000000 -0500
@@ -178,3 +178,5 @@
 	.align 3
 .LEFDE1:
 #endif
+
+.section .note.GNU-stack,"",@progbits	
--- Python-2.5/Modules/_ctypes/libffi/src/powerpc/linux64_closure.S.ctypes-exec	2007-01-03 12:54:40.000000000 -0500
+++ Python-2.5/Modules/_ctypes/libffi/src/powerpc/linux64_closure.S	2007-01-03 12:53:51.000000000 -0500
@@ -204,3 +204,5 @@
 	.align 3
 .LEFDE1:
 #endif
+
+.section .note.GNU-stack,"",@progbits		
--- Python-2.5/Modules/_ctypes/libffi/src/powerpc/ppc_closure.S.ctypes-exec	2007-01-03 12:54:40.000000000 -0500
+++ Python-2.5/Modules/_ctypes/libffi/src/powerpc/ppc_closure.S	2007-01-03 12:52:35.000000000 -0500
@@ -321,3 +321,5 @@
 .LEFDE1:
 
 #endif
+
+.section .note.GNU-stack,"",@progbits	
--- Python-2.5/Modules/_ctypes/libffi/src/powerpc/sysv.S.ctypes-exec	2007-01-03 12:54:40.000000000 -0500
+++ Python-2.5/Modules/_ctypes/libffi/src/powerpc/sysv.S	2007-01-03 12:51:57.000000000 -0500
@@ -215,3 +215,5 @@
       .align 2
 .LEFDE1:
 #endif
+
+.section .note.GNU-stack,"",@progbits	
--- Python-2.5/Modules/_ctypes/libffi/src/ia64/unix.S.ctypes-exec	2007-01-03 12:54:40.000000000 -0500
+++ Python-2.5/Modules/_ctypes/libffi/src/ia64/unix.S	2007-01-03 12:54:05.000000000 -0500
@@ -553,3 +553,5 @@
 	data8	@pcrel(.Lld_hfa_float)		// FFI_IA64_TYPE_HFA_FLOAT
 	data8	@pcrel(.Lld_hfa_double)		// FFI_IA64_TYPE_HFA_DOUBLE
 	data8	@pcrel(.Lld_hfa_ldouble)	// FFI_IA64_TYPE_HFA_LDOUBLE
+
+.section .note.GNU-stack,"",@progbits			
--- Python-2.5/Modules/_ctypes/libffi/src/sparc/v8.S.ctypes-exec	2007-01-03 12:54:40.000000000 -0500
+++ Python-2.5/Modules/_ctypes/libffi/src/sparc/v8.S	2007-01-03 12:54:24.000000000 -0500
@@ -265,3 +265,5 @@
 	.byte	0x1f	! uleb128 0x1f
 	.align	WS
 .LLEFDE2:
+
+.section .note.GNU-stack,"",@progbits	
--- Python-2.5/Modules/_ctypes/libffi/src/sparc/v9.S.ctypes-exec	2007-01-03 12:54:40.000000000 -0500
+++ Python-2.5/Modules/_ctypes/libffi/src/sparc/v9.S	2007-01-03 12:54:26.000000000 -0500
@@ -300,3 +300,5 @@
 	.align 8
 .LLEFDE2:
 #endif
+
+.section .note.GNU-stack,"",@progbits	
