--- Python-2.5-orig/Modules/_tkinter.c	2006-08-11 22:33:36.000000000 -0400
+++ Python-2.5/Modules/_tkinter.c	2007-10-19 01:04:42.000000000 -0400
@@ -938,7 +938,7 @@
 #if defined(Py_UNICODE_WIDE) && TCL_UTF_MAX == 3
 		Tcl_UniChar *outbuf;
 		Py_ssize_t i;
-		assert(size < size * sizeof(Tcl_UniChar));
+		assert(size == 0 || size < size * sizeof(Tcl_UniChar));
 		outbuf = (Tcl_UniChar*)ckalloc(size * sizeof(Tcl_UniChar));
 		if (!outbuf) {
 			PyErr_NoMemory();
