--- Python-2.5/Lib/xmlrpclib.py.orig	2007-04-10 10:29:14.000000000 -0400
+++ Python-2.5/Lib/xmlrpclib.py	2007-06-19 12:08:04.000000000 -0400
@@ -630,6 +630,9 @@
         try:
             f = self.dispatch[type(value)]
         except KeyError:
-            raise TypeError, "cannot marshal %s objects" % type(value)
+            if isinstance(value, object):
+                self.dump_instance(value, write)
+            else:
+                raise TypeError, "cannot marshal %s objects" % type(value)
         else:
             f(self, value, write)
